src/nccl_impl.cpp

```diff
@@ -57,8 +44,12 @@ void NCCLCommunicator::nccl_setup() {
   AL_CHECK_NCCL(ncclCommInitRank(&m_nccl_comm, size(), nccl_id, rank()));
 }
 
-void NCCLCommunicator::nccl_destroy() {
-  AL_CHECK_NCCL(ncclCommDestroy(m_nccl_comm));
+NCCLCommunicator::~NCCLCommunicator() {
+  int d;
+  // Only destroy resources if the driver is still loaded.
+  if (cudaGetDevice(&d) == cudaSuccess) {
+    AL_CHECK_NCCL(ncclCommDestroy(m_nccl_comm));
```

Comment:
Micro-optimization: [avoid throwing exceptions in destructors](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c36-a-destructor-may-not-fail):
```suggestion
    ncclCommDestroy(m_nccl_comm);
```

Reply:
I disagree with this suggestion. If the `ncclCommDestroy` fails, the process should terminate.

Reply:
The "proper" fix is to either call terminate yourself or just declare the destructor `noexcept` and C++ will take care of that for you... (You'll have to turn off the compiler warning then, though.)

Reply:
I will fix it properly.